aclk/agent_cloud_link.c

```diff
@@ -327,59 +403,295 @@ struct aclk_query *aclk_queue_pop()
+static void _dump_connector_list()
+{
+
+    struct _collector  *tmp_collector;
 
-        ACLK_UNLOCK;
+    COLLECTOR_LOCK;
 
+    info("DUMPING ALL COLLECTORS");
+
+    if (unlikely(!collector_list || !collector_list->next)) {
+        COLLECTOR_UNLOCK;
+        info("DUMPING ALL COLLECTORS -- nothing found");
+        return;
+    }
+
+    // Note that the first entry is "dummy"
+    tmp_collector = collector_list->next;
+
+    while (tmp_collector) {
+        info(
+            "COLLECTOR %s : [%s:%s] count = %u", tmp_collector->hostname,
+            tmp_collector->plugin_name ? tmp_collector->plugin_name : "",
+            tmp_collector->module_name ? tmp_collector->module_name : "", tmp_collector->count);
+
+        tmp_collector = tmp_collector->next;
+
+    }
+    info("DUMPING ALL COLLECTORS DONE");
+    COLLECTOR_UNLOCK;
+}
+#endif
+
+/*
+ * This will cleanup the collector list
+ *
+ */
+static void _reset_connector_list()
+{
+    struct _collector  *tmp_collector, *next_collector;
+
+    COLLECTOR_LOCK;
+
+    if (unlikely(!collector_list || !collector_list->next)) {
+        COLLECTOR_UNLOCK;
+        return;
+    }
+
+    // Note that the first entry is "dummy"
+    tmp_collector  = collector_list->next;
+    collector_list->count = 0;
+    collector_list->next  = NULL;
+
+    // We broke the link; we can unlock
+    COLLECTOR_UNLOCK;
+
+    while (tmp_collector) {
+        next_collector = tmp_collector->next;
+        _free_collector(tmp_collector);
+        tmp_collector = next_collector;
+    }
+}
+
+
+/*
+ * Find a collector (if it exists)
+ * Must lock before calling this
+ * If last_collector is not null, it will return the previous collector in the linked
+ * list (used in collector delete)
+ */
+static struct _collector *_find_collector(const char *hostname, const char *plugin_name, const char *module_name, struct _collector **last_collector)
+{
+    struct _collector *tmp_collector, *prev_collector;
+    uint32_t plugin_hash;
+    uint32_t module_hash;
+    uint32_t hostname_hash;
+
+    if (unlikely(!collector_list)) {
+        collector_list = callocz(1, sizeof(struct _collector));
         return NULL;
     }
 
-    if (unlikely(action == PUBLICH_TOPIC_REBUILD)) {
-        ACLK_UNLOCK;
-        get_publish_base_topic(PUBLICH_TOPIC_FREE);
-        return get_publish_base_topic(PUBLICH_TOPIC_GET);
+    if (unlikely(!collector_list->next))
+        return NULL;
+
+    plugin_hash = plugin_name?simple_hash(plugin_name):1;
+    module_hash = module_name?simple_hash(module_name):1;
+    hostname_hash = simple_hash(hostname);
+
+    // Note that the first entry is "dummy"
+    tmp_collector  = collector_list->next;
+    prev_collector = collector_list;
+    while (tmp_collector) {
+        if (plugin_hash == tmp_collector->plugin_hash &&
+            module_hash == tmp_collector->module_hash &&
+            hostname_hash == tmp_collector->hostname_hash &&
+            (!strcmp(hostname, tmp_collector->hostname)) &&
```

Comment:
why compare hostname by hash and then also full string compare?

Reply:
That's the original code from the netdata agent that Kosta wrote. The idea is that if the hashes are different you did a very fast comparison.

Reply:
ok thanks I get it, it's faster in case when this condition is evaluated where most of the time those strings are different. Why not use the same logic then for plugin_hash and module_hash?

Reply:
ok thanks, clear